less_users/views.py

```diff
@@ -40,104 +48,231 @@ class ProfileView(LoginRequiredMixin, View):
-    """gain points when challenge is completed; stop challenge or detele challenge """
-    user_id = request.user
-    if 'done' in request.POST:
-        """should check if points can be added (depending on durration and frequncy in model"""
-        challenge_id = int(request.POST.get('done'))
-        points = Challenge.objects.get(id=challenge_id).points
-        this = UserChallenge.objects.get(user_id=user_id, challenge_id=challenge_id, is_active=True)
-        user_challenge_id = this.id
-        this.check_if_active()
-        if this.get_points == points and this.is_active:
-            new_log = Log.objects.create(user_challenge_id=user_challenge_id, points=points)
-            my_profile = Profile.objects.get(user_id=user_id)
-            my_profile.points += points
-            my_profile.save()
-            messages.success(request, f'You have got new points {points} for challenge: {this.challenge.name}!')
-        elif this.get_points == 0 or this.is_active == False:
-            messages.warning(request, f'You cannot get new points for this challenge yet: {this.challenge.name}!')
-    elif 'stop' in request.POST:
-        user_challenge_id = int(request.POST.get('stop'))
-        this = UserChallenge.objects.get(id=user_challenge_id)
-        this.is_active = False
-        this.save()
-        messages.info(request, f'You stopped challenge: {this.challenge.name}. You can activate the new one.')
-    elif 'delete' in request.POST:
-        user_challenge_id = int(request.POST.get('delete'))
-        this = UserChallenge.objects.get(id=user_challenge_id)
-        this.delete()
-        messages.warning(request, f'You have deleted {this.challenge.name} from your challenges')
-    return redirect('my_challenges')
+        UserChallenge.objects.create(user=request.user, challenge_id=pk)
+        messages.success(request, "You activated new challenge!")
+        return HttpResponseRedirect(reverse("challenge_detail", args=[str(pk)]))
+
+
+def event_view_done(request):
+    """gain points when user challenge is completed by clicking >done< button;
+    should check if points can be added (depending on duration and frequency in model)"""
+    try:
+        challenge_id = int(request.POST.get("done"))
+    except ValueError:
+        messages.warning(request, "This challenge does not exist")
+    else:
+        if Challenge.objects.filter(pk=challenge_id).exists():
+            points = Challenge.objects.get(pk=challenge_id).points
+            # check also not active user_challenges and their logs for frequency!
+            if UserChallenge.objects.filter(
+                user_id=request.user, challenge_id=challenge_id
+            ).exists():
+                user_challenge = UserChallenge.objects.get(
+                    user_id=request.user,
+                    challenge_id=challenge_id,
+                    is_active=True,
+                    is_deleted=False,
+                )
+                if user_challenge:
```

Comment:
Masz dwa zapytania do bazy i jeszcze dwa razy sprawdzasz to samo

Reply:
ojoj, tu w ogóle się dzieje... dużo niedobrego i powtarzalnego :/ jeszcze przetwarzam...